### ISSUE TITLE
await-promise is incompatible with the Promise/A+ spec

### ISSUE THREAD

**Issue by alexandru** [palantir/tslint#3512]
### Bug Report

- __TSLint version__: 5.8.0
- __TypeScript version__: 2.6.1

#### TypeScript code being linted

```ts
  const tweetsIterator = twitterClient.scanUserTimeline(account.twitter)
  while (true) {
    const tweets = await tweetsIterator.next()
    if (tweets.value && tweets.value.length) {
      await db.bulkInsertTweets(twitterArchiveIndex, tweets.value)
    }
    if (tweets.done) {
      break
    }
  }
```

#### Actual behavior

```
Invalid 'await' of a non-Promise value. (await-promise)
  68 |       const tweetsIterator = twitterClient.scanUserTimeline(account.twitter)
  69 |       while (true) {
> 70 |         const tweets = await tweetsIterator.next()
     |                       ^
  71 |         if (tweets.value && tweets.value.length) {
  72 |           await db.bulkInsertTweets(twitterArchiveIndex, tweets.value)
  73 |         }
```

#### Expected behavior

The error is triggered because the object awaited is a [Funfix Future](https://funfix.org/api/exec/classes/future.html) which does not inherit from `Promise`.

According to the [Promises/A+](https://promisesaplus.com/) specification the `Promise` class is NOT required for `await` to work. You only need a "thenable" object, meaning a type that implements `then` with the described behavior.

In TypeScript this is exposed as `PromiseLike`:

```ts
interface PromiseLike<T> {
    /**
     * Attaches callbacks for the resolution and/or rejection of the Promise.
     * @param onfulfilled The callback to execute when the Promise is resolved.
     * @param onrejected The callback to execute when the Promise is rejected.
     * @returns A Promise for the completion of which ever callback is executed.
     */
    then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): PromiseLike<TResult1 | TResult2>;
}
```

Possibly related: https://github.com/palantir/tslint/issues/3381

Please make this check look for `PromiseLike` and not for `Promise`.

Thanks,

**Comment by ajafff** [comment]
You can configure what the rule should treat as promise: https://palantir.github.io/tslint/rules/await-promise/

**Comment by alexandru** [comment]
What's the motivation for `Thenable` to not e the default?

**Comment by ajafff** [comment]
The rule doesn't check for assignability / compatibility. It just checks the name. `Thenable` is a builtin type and even if it was, you are not required to implement that interface.

**Comment by alexandru** [comment]
You mean it literally checks that the type's name is that string?

**Comment by ajafff** [comment]
> You mean it literally checks that the type's name is that string?

Yep, it checks the name of every base type and recurses into union and intersection types.

**Comment by adidahiya** [comment]
How's this for a docs update to clarify the issue? https://github.com/palantir/tslint/pull/3519
